### PR TITLE
Add image classification models

### LINKED ISSUE #28: Example configuration for image classification.
In this example, we would like to show how to do image classification task in PaddlePaddle. We need to provide several classic configuration, AlexNet, VGG, GoogleNet, ResNet. It's betther to provide GoogleNet-v3/v4, [identity-mapping-ResNet](https://arxiv.org/abs/1603.05027), but it doesn't matter if they are not provided in the first version.

Reference configuration:

- v1 of AlexNet: https://github.com/PaddlePaddle/Paddle/blob/develop/benchmark/paddle/image/alexnet.py
- VGG: https://github.com/PaddlePaddle/book/blob/develop/03.image_classification/vgg.py
- v1 of GoogleNet: https://github.com/PaddlePaddle/Paddle/blob/develop/benchmark/paddle/image/googlenet.py
- ResNet: https://github.com/PaddlePaddle/book/blob/develop/03.image_classification/resnet.py

DataSet:
- ImageNet
- Oxford-flowers: http://www.robots.ox.ac.uk/~vgg/data/flowers/102/index.html

### PR DESCRIPTION
**Pull request by wwhu**
resolve #28 
The classification accuracy has not been validated yet.

### CODE REVIEW

File: image_classification/googlenet.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
+import paddle.v2 as paddle
+
+__all__ = ['googlenet']
+
+
+def inception(name, input, channels, filter1, filter3R, filter3, filter5R,
+              filter5, proj):
+    cov1 = paddle.layer.conv_projection(
```
Comment (qingqing01) [inline image_classification/googlenet.py]: 这里采用这个配置里inception2 的写法吧，https://github.com/PaddlePaddle/Paddle/blob/develop/benchmark/paddle/image/googlenet.py#L19  

`conv_projection`不适应CPU.

File: image_classification/resnet.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        152: ([3, 8, 36, 3], bottleneck)
+    }
+    stages, block_func = cfg[depth]
+    conv1 = conv_bn_layer(
+        input, ch_in=3, ch_out=64, filter_size=7, stride=2, padding=3)
+    pool1 = paddle.layer.img_pool(input=conv1, pool_size=3, stride=2)
+    res1 = layer_warp(block_func, pool1, 64, stages[0], 1)
+    res2 = layer_warp(block_func, res1, 128, stages[1], 2)
+    res3 = layer_warp(block_func, res2, 256, stages[2], 2)
+    res4 = layer_warp(block_func, res3, 512, stages[3], 2)
```
Comment (qingqing01) [inline image_classification/resnet.py]: 对于上面  " # TODO: bug fix for ch_in = input.num_filters "，一种办法是这里`layer_warp`这里可以显示的指定通道数。 当然也可以等配置解析重写之后，看能不能获取layer的属性，再改也行~~

File: image_classification/train.py [hunk 3]
```diff
@@ -45,16 +70,19 @@ def main():
...
-        learning_rate=0.01 / BATCH_SIZE,
+        learning_rate=0.001 / BATCH_SIZE,
         learning_rate_decay_a=0.1,
         learning_rate_decay_b=128000 * 35,
         learning_rate_schedule="discexp", )
 
     train_reader = paddle.batch(
-        paddle.reader.shuffle(reader.test_reader("train.list"), buf_size=1000),
+        paddle.reader.shuffle(
+            reader.test_reader(os.path.join(args.data_dir, 'train.list')),
```
Comment (qingqing01) [inline image_classification/train.py]: 等haoshuang的[PR](https://github.com/PaddlePaddle/Paddle/pull/2242/files)合入之后，需要换成flowers的数据，用加速的reader。 同时文档里，可以告诉用户如果换成自己的数据集如何处理~

File: image_classification/train.py [hunk 4]
```diff
@@ -1,38 +1,63 @@
...
+            input=fc_o1, size=CLASS_DIM, act=paddle.activation.Softmax())
+        loss1 = paddle.layer.cross_entropy_cost(
+            input=out1, label=lbl, coeff=0.3)
+        paddle.evaluator.classification_error(input=out1, label=lbl)
+        out2 = paddle.layer.fc(
+            input=fc_o2, size=CLASS_DIM, act=paddle.activation.Softmax())
+        loss2 = paddle.layer.cross_entropy_cost(
+            input=out2, label=lbl, coeff=0.3)
+        paddle.evaluator.classification_error(input=out2, label=lbl)
+        extra_layers = [loss1, loss2]
```
Comment (qingqing01) [inline image_classification/train.py]: 要么把net, cost这些代码都放到各自的配置里去？这样train.py看着清爽一些~

**Review by qingqing01 — COMMENTED** [review]
还需要使用文档，可以参考其他models里的文档，但模型介绍部分觉得不用详细讲了，都是比较经典的模型，book里都有介绍，可以指向链接指向book。

File: image_classification/googlenet.py (hunk 1 above)
Comment (wwhu) [reply image_classification/googlenet.py]: 已添加inception2

File: image_classification/resnet.py (hunk 2 above)
Comment (wwhu) [reply image_classification/resnet.py]: 输入通道数的用处是判断是否需要在block之间进行projection，我这里直接通过b_projection参数显式指定是否需要projection

File: image_classification/train.py (hunk 3 above)
Comment (wwhu) [reply image_classification/train.py]: 已在文档里说明如何表示数据

File: image_classification/train.py (hunk 4 above)
Comment (wwhu) [reply image_classification/train.py]: 把net放到各自配置里了，因为考虑到infer阶段不需要cost，所以cost还是放到了train.py

File: image_classification/googlenet.py (hunk 1 above)
Comment (qingqing01) [reply image_classification/googlenet.py]: 直接去掉inception， 只保留inception2吧~

File: image_classification/README.md [hunk 5]
```diff
@@ -1 +1,182 @@
...
+parameters = paddle.parameters.create(cost)
+
+# Create optimizer
+optimizer = paddle.optimizer.Momentum(
+    momentum=0.9,
+    regularization=paddle.optimizer.L2Regularization(rate=0.0005 *
+                                                     BATCH_SIZE),
+    learning_rate=0.001 / BATCH_SIZE,
+    learning_rate_decay_a=0.1,
+    learning_rate_decay_b=128000 * 35,
```
Comment (qingqing01) [inline image_classification/README.md:108]: learning_rate_decay_a， learning_rate_decay_b解释下吧，参见book里的解释~

File: image_classification/resnet.py (hunk 2 above)
Comment (qingqing01) [reply image_classification/resnet.py]: 明白了，这样也可以，那上面的 TODO 可以删掉了吧~

**Comment by wwhu** [comment]
@qingqing01 这几处已修改

File: image_classification/resnet.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    conv2 = conv_bn_layer(conv1, ch_out, 3, 1, 1)
+    conv3 = conv_bn_layer(conv2, ch_out * 4, 1, 1, 0,
+                          paddle.activation.Linear())
+    short = shortcut(input, ch_out * 4, stride, b_projection)
+    return paddle.layer.addto(
+        input=[conv3, short], act=paddle.activation.Relu())
+
+
+def layer_warp(block_func, input, features, count, stride):
+    conv = block_func(input, features, stride, True)
```
Comment (qingqing01) [inline image_classification/resnet.py]: 第一个block_func的`b_projection`不是总为`True`的，还是的依赖于输入conv的channel数。 所以我觉得还是`shortcut`里对判断输入和输出的channel数是否一致，决定是否用conv做升降维度好些~

**Review by qingqing01 — COMMENTED** [review]
再加个`infer.py`吧。

**Comment by qingqing01** [comment]
flowers dataset已经merge了，需要换下~

**Comment by guoshengCS** [comment]
能否稍微调整下`basicblock`和`bottleneck`中`shortcut`的顺序，将`shortcut`放在前面，保证拓扑顺序与原文尽可能一致，thx~

File: image_classification/resnet.py (hunk 6 above)
Comment (wwhu) [reply image_classification/resnet.py]: 已修改

**Comment by wwhu** [comment]
添加了`infer.py`以及flowers数据，并调整了`shortcut`位置。
@qingqing01 @guoshengCS

File: image_classification/infer.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        # The storage order of the loaded image is W(widht),
+        # H(height), C(channel). PaddlePaddle requires
+        # the CHW order, so transpose them.
+        im = im.transpose((2, 0, 1))  # CHW
+        # In the training phase, the channel order of CIFAR
+        # image is B(Blue), G(green), R(Red). But PIL open
+        # image in RGB mode. It must swap the channel order.
+        im = im[(2, 1, 0), :, :]  # BGR
+        im = im.flatten()
+        im = im / 255.0
```
Comment (qingqing01) [inline image_classification/infer.py]: 这里看下image.py里有没有函数可以直接用。而且这里 `im/255.0`和训练不对应吧，训练是减去均值吧。

File: image_classification/googlenet.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
+import paddle.v2 as paddle
+
+__all__ = ['googlenet']
+
+
+def inception2(name, input, channels, filter1, filter3R, filter3, filter5R,
```
Comment (qingqing01) [inline image_classification/googlenet.py]: inception2 -> inception

File: image_classification/vgg.py [hunk 9]
```diff
@@ -17,7 +17,7 @@
 __all__ = ['vgg13', 'vgg16', 'vgg19']
 
 
-def vgg(input, nums):
+def vgg(input, nums, class_dim=100):
```
Comment (qingqing01) [inline image_classification/vgg.py]: class_dim=100 ->  class_dim，去掉100的默认值吧

File: image_classification/resnet.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+
+def layer_warp(block_func, input, ch_in, ch_out, count, stride):
+    conv = block_func(input, ch_in, ch_out, stride)
+    for i in range(1, count):
+        conv = block_func(conv, ch_out, ch_out, 1)
+    return conv
+
+
+def resnet_imagenet(input, depth=50, class_dim=100):
```
Comment (qingqing01) [inline image_classification/resnet.py]: class_dim=100 -> class_dim， 去掉默认值100吧。

**Comment by lcy-seso** [comment]
If you are not going to finish this work. please tell me.

**Comment by wwhu** [comment]
I will finish it ASAP. Sorry for the delay. @lcy-seso

**Comment by lcy-seso** [comment]
@wwhu  You are welcome. I think after some modifications, we can try to merge the already finished part and then based on a merged version, refactor, refine, and validate the learning performance. Also, thanks for your work.

File: image_classification/infer.py (hunk 7 above)
Comment (wwhu) [reply image_classification/infer.py]: 已修改

File: image_classification/resnet.py (hunk 10 above)
Comment (wwhu) [reply image_classification/resnet.py]: 所有的默认值都去掉了

**Review by qingqing01 — APPROVED** [review]
LGTM.  精度需要进一步确保~
